### PR TITLE
List.equal, List.compare

### PR DESCRIPTION
**Pull request by gasche**
`List.equal f foo bar` is nicer than
`List.length foo = List.length bar && List.for_all2 f foo bar`.

Note: with List.compare there is a risk of users having opened the
List module, and then using 'compare' from the stdlib unqualified. For
example:

    List.(sort compare foo bar)

Such code will break (type error), and has to be fixed by using
Stdlib.compare. Stdlib is available since OCaml 4.07; people wishing
to support both 4.12 and older releases would have to avoid opening
List, or rebind 'compare' locally.

### CODE REVIEW

**Comment by gasche** [comment]
Notes:
- the proposed functions are exactly consistent with those of `Map.S` (including the fact of not checking sizes before checking element-wise equality)
- #84 added a bunch of equal functions for non-parametrized types (Bytes, Char, Int32, Int64, Nativeint and String).

File: stdlib/list.mli
Comment (nojb) [inline stdlib/list.mli:110]: Nitpicking: `at the same position` superfluous given the notation.

**Review by nojb — APPROVED** [review]
LGTM! Clearly useful functions, and obvious implementation.

(removed an additional comment that was nonesense)

**Comment by lthls** [comment]
I'm curious to know if compatibility between `List.compare Stdlib.compare` and `Stdlib.compare` (on lists) can be expected, happened by pure chance, or should not be relied upon.

**Comment by gasche** [comment]
@lthls for, say, tuples and arrays, I'm in the camp that doesn't want to rely on `Stdlib.compare` being a lexicographic order. For lists, given their inductive structure, I think that there is exactly one way to compare lists that feels right, so I would say that it is an expected coincidence that `Stdlib.compare` and `List.compare Stdlib.compare` coincide.

File: stdlib/list.mli
Comment (gasche) [reply stdlib/list.mli:110]: I think that having the mention may make things more accessible for people that are less familiar with mathematical notation.

**Comment by gasche** [comment]
This PR was approved by one maintainer, but I think the norm for stdlib PRs is to get more eyeballs and prefer a second approval. Would someone be willing to have a look and give an opinion?

Note: on bad days I wonder if `List.equal : ('a -> 'a -> bool) -> 'a list -> 'a list -> bool` should be called `List.related` to emphasize that non-equivalence relations can also be used here. I think `List.equal (<)` (to check that two lists have pairwise-ordered elements) could be a bit confusing. (But this could be improved later with an alias?)

File: stdlib/list.mli
Comment (dbuenzli) [inline stdlib/list.mli:108]: `..` -> `...`

File: stdlib/list.mli
Comment (dbuenzli) [reply stdlib/list.mli:110]: On the other hand it makes it harder to understand/more confusing for those who know. Drawing from other doc strings below you could say:
```
[equal eq [a1; ...; an] [b1; ...; bn]] is [eq a1 b1 && eq a2 b2 && ... && eq an bn] or 
[false] if the lists are not of the same size. 
```

File: stdlib/list.mli
Comment (dbuenzli) [inline stdlib/list.mli:132]: I understand the new generation of programmers tends to be averse to learning but that's kind of implied by the "lexicographic" mentioned above. It's needless wordy in my opinion.

File: stdlib/list.mli
Comment (dbuenzli) [inline stdlib/list.mli]: The `(negative result)` reads weird here. I scratched my head. I would rather say: `The result is negative if ... is smaller than ..., that is`

File: stdlib/list.mli
Comment (gasche) [reply stdlib/list.mli:110]: Great suggestion, thanks.

**Review by dbuenzli — APPROVED** [review]
Left a few non-essential doc strings nitpicks.

**Comment by gasche** [comment]
I'm getting tired of waiting for a second maintainer to approve this PR. I will integrate some of @dbuenzli's nitpicks, and then consider that his approval in addition to @nojb's is good enough (getting @dbuenzli's approval is a fairly high standard for stdlib PRs).

File: stdlib/list.mli
Comment (gasche) [reply stdlib/list.mli]: Thanks for the good feedback. I didn't find a nice formulation to at the same time explain lexicographic ordering (yes) and explain the integer convention, so I just dropped the second part -- hopefully the reference to `Stdlib.compare` is enough.

File: stdlib/list.mli
Comment (gasche) [reply stdlib/list.mli:132]: I kept the wordy explanation.

**Comment by xavierleroy** [comment]
Two months later, should we merge?

File: stdlib/list.ml
Comment (alainfrisch) [inline stdlib/list.ml:556]:
```suggestion
   starting with distinct elements. It is also incorrect for
```

File: stdlib/list.mli
Comment (alainfrisch) [inline stdlib/list.mli:107]: What about:
````ocaml
val equal : ('a -> 'b -> bool) -> 'a list -> 'b list -> bool
````
This might be a bit weird for an equality function, but it also makes it more useful (and it makes it clear that it's just a variant of `for_all2`).

File: stdlib/list.mli
Comment (alainfrisch) [inline stdlib/list.mli:121]: Same here: is there a strong reason to force the same type on the two lists?

**Comment by gasche** [comment]
@alainfrisch mixed feelings about your proposal to work with heterogeneous relations `('a -> 'b -> {bool,int})`:
- it's more generality for free
- but it's inconsistent with the type of equal/compare in `Map` and `Option`
  (maybe we should also generalize those?)

Given that we can allow ourselves to generalize the interface later, I would be in favor of using the homogeneous versions for now, and, if we want to generalize, also change Map and Option.

**Comment by alainfrisch** [comment]
> Given that we can allow ourselves to generalize the interface later, I would be in favor of using the homogeneous versions for now, and, if we want to generalize, also change Map and Option.

Fair enough!

**Review by alainfrisch — APPROVED** [review]

**Comment by gasche** [comment]
I rebased to fix the Changes conflict, anyone should feel free to merge if the CI is happy again. (I wouldn't be entirely surprised if some testsuite noise needed another round of fix.)

**Comment by xavierleroy** [comment]
AppVeyor seems stuck but the other two CI are green, so I'll merge as instructed.

File: stdlib/list.ml
Comment (iguerNL) [inline stdlib/list.ml:575]: @gasche I've just seen the addition of these two functions (on discuss). Depending of the usage contexts and how lists are constructed, I was wondering if it's not a good idea to add pre-comparison l1 == l2 to immediately answer when the two lists are physically equal.

Comment (jberdine) [reply stdlib/list.ml:575]: Agree. Maybe the code I interact with is odd, but the profiler has taught me to always start structural comparison functions like these with a physical equality test. If it doesn't help the cost is usually nothing and at worst very small. But on the other hand the savings on inputs with some internal sharing can be very significant.

Comment (gasche) [reply stdlib/list.ml:575]: This optimization would only be valid if the `cmp` parameter is reflexive (it always returns `true` on equal inputs), for example `List.equal (fun _ _ -> false) [()] [()]` changes behavior. This makes me inclined to think that we may not want this trick. (As far as I know, the polymorphic-equality usage in the standard library are always valid, they do not require extra assumptions.)

Comment (jberdine) [reply stdlib/list.ml:575]: Ah, ok, fair point. I'm implicitly assuming that the compare functions should be at least preorders. I'm not familiar with / accustomed to uses where they are not reflexive.

Comment (iguerNL) [reply stdlib/list.ml:575]: Ok, I understand.

Comment (gasche) [reply stdlib/list.ml:575]: One can use `List.equal` as a variant of `List.for_all2` that behaves nicely on different-length lists, and then the predicate may be arbitrary. I don't really like this usage either, but I don't have a much better proposal to offer for those use-cases.

Comment (lthls) [reply stdlib/list.ml:575]: > This optimization would only be valid if the `cmp` parameter is reflexive (it always returns `true` on equal inputs), for example `List.compare (fun _ _ -> false) [()] [()]` changes behavior. This makes me inclined to think that we may not want this trick. (As far as I know, the polymorphic-equality usage in the standard library are always valid, they do not require extra assumptions.)

I can understand the argument for the `equal` function, but I don't think it applies as well to `compare`. Given that it is explicitly documented that `x == y` implies `Stdlib.compare x y = 0`, I would expect other `compare` functions from the standard library to provide similar guarantees.
(If I remember correctly, with custom values you can get cases where `Stdlib.compare x x` would have returned a non-zero result if the physical equality shortcut didn't trigger, so the fact that `List.compare` is taking a function as argument, and that argument is not guaranteed to be a total order, is not necessarily a fundamental difference with `Stdlib.compare`).
On the other hand, this does change the semantics for comparison on infinite lists. If I was asked what should be the semantics of `let rec l = 0 :: l in List.compare Stdlib.compare l l` I would be fine with both returning `0` and looping forever. If you prefer the looping version that would make a decent argument for keeping `compare` as it is.

Comment (gasche) [reply stdlib/list.ml:575]: I agree that it makes more sense for `compare`. (The snippet I wrote used `compare` but it was a mistake, I rewrote it with `equal` as originally intended.) Maybe you want to submit a PR with this argument?

Comment (jberdine) [reply stdlib/list.ml:575]: Done: https://github.com/ocaml/ocaml/pull/10239

**Pull request merged**
